src/Stratis.Bitcoin/Utilities/MemoryCache.cs

```diff
@@ -0,0 +1,134 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Stratis.Bitcoin.Utilities
+{
+    /// <summary>
+    /// Memory cache that implements the Least Recently Used (LRU) standard.
+    /// </summary>
+    public class MemoryCache<TKey, TValue>
+    {
+        /// <summary>Dictionary that contains cached items.</summary>
+        private readonly Dictionary<TKey, LinkedListNode<CacheItem<TKey, TValue>>> cache;
+
+        /// <summary>Keys sorted by their last access time with most recent ones at the end.</summary>
+        private readonly LinkedList<CacheItem<TKey, TValue>> keys;
+
+        /// <summary>Maximum items count that can be stored in the cache.</summary>
+        private readonly int maxItemsCount;
+        
+        /// <summary>Lock to protect access to <see cref="keys"/> and <see cref="cache"/>.</summary>
```

Comment:
mention these contracts in remark comments of dictionary and linked list as well

Reply:
Fixed